### PR TITLE
Decrement stream counts immediately when the stream is removed from stream_list

### PR DESCRIPTION
**Pull request by masaori335**
HTTP/2 active stream counts (`client_streams_in_count` and `client_streams_out_count`) are decremented in `Http2Stream::release_stream()`. This is introduced by dc8be341578aa0fa900c84265d0a336414dfad9b. But these counts should be decremented just after the stream is removed from `stream_list` in `Http2Stream::delete_stream()` like these counts are incremented just after the stream is added in `stream_list`. Because when ATS sends END_STREAM flag or RST_STREAM frame, endpoints might open a new stream.

----

- Fix a issue of #3100
- 7.1.x backport candidate

### CODE REVIEW

**Comment by masaori335** [comment]
I tested this on 7.1.x with same condition of #3100, the failures are gone. ( I can't test master because of #3101. )

```
$ h2load -n 10000 -m 100 -c 2 https://127.0.0.1:4443/100k
...
finished in 7.17s, 1394.70 req/s, 136.31MB/s
requests: 10000 total, 10000 started, 10000 done, 10000 succeeded, 0 failed, 0 errored, 0 timeout
status codes: 10000 2xx, 0 3xx, 0 4xx, 0 5xx
...
```

**Comment by masaori335** [comment]
[approve ci icc]

File: proxy/http2/Http2ConnectionState.cc [hunk 1]
```diff
@@ -1160,16 +1170,8 @@ Http2ConnectionState::delete_stream(Http2Stream *stream)
 void
 Http2ConnectionState::release_stream(Http2Stream *stream)
 {
-  // Update stream counts
   if (stream) {
     --total_client_streams_count;
```
Comment (maskit) [inline proxy/http2/Http2ConnectionState.cc]: Leave some comment why only `total_client_streams_count` need to be decremented here and others are not.

**Comment by zwoop** [comment]
[approve ci icc] 
[approve ci autest]

**Comment by zwoop** [comment]
Cherry-picked to 7.1.x

**Pull request merged**